<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>pctax: Analyzing Omics Data with R - 9&nbsp; Enrichment analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./10-element-cycle.html" rel="next">
<link href="./08-network.html" rel="prev">
<link href="./genomics.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script>
  MathJax = {
    tex: {
      inlineMath: [['$', '$'], ['\\(', '\\)']]
    }
  };
</script><script id="MathJax-script" async="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script><script src="site_libs/kePrint-0.0.1/kePrint.js"></script><link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="site_libs/htmlwidgets-1.6.1/htmlwidgets.js"></script><script src="site_libs/sankeyNetwork-binding-0.3.2/sankeyNetwork.js"></script><script src="site_libs/d3-4.13.0/d3.min.js"></script><script src="site_libs/sankey-0.1/sankey.js"></script>
<meta property="og:title" content="pctax: Analyzing Omics Data with R - 9&nbsp; Enrichment analysis">
<meta property="og:description" content="">
<meta property="og:site_name" content="pctax: Analyzing Omics Data with R">
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./09-enrichment.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Enrichment analysis</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">pctax: Analyzing Omics Data with R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/asa12138/pctax_tutorial" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Example data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-metadata.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Display metadata</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-composition.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Explore composition</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Diversity analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-differential.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Differential abundance analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-cluster.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Clustering analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community assembly</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08-network.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Network analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09-enrichment.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Enrichment analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-element-cycle.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Elemental cycling</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-others.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Other features</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#reporterscore" id="toc-reporterscore" class="nav-link active" data-scroll-target="#reporterscore"><span class="header-section-number">9.1</span> ReporterScore</a></li>
  <li><a href="#kegg" id="toc-kegg" class="nav-link" data-scroll-target="#kegg"><span class="header-section-number">9.2</span> KEGG</a></li>
  <li><a href="#args-vfs" id="toc-args-vfs" class="nav-link" data-scroll-target="#args-vfs"><span class="header-section-number">9.3</span> ARGs &amp; VFs</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/asa12138/pctax_tutorial/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/asa12138/pctax_tutorial/blob/main/09-enrichment.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Enrichment analysis</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>Assess the functional significance of your omics data through integrated functional enrichment analysis using <a href="https://github.com/Asa12138/ReporterScore">ReporterScore</a>. Understand the biological implications of detected features in the context of known pathways and processes.</p>
<p>For more details, refer to ‘C. Peng, Q. Chen, S. Tan, X. Shen, C. Jiang, Generalized Reporter Score-based Enrichment Analysis for Omics Data. bioRxiv [Preprint] (2023). <a href="https://doi.org/10.1101/2023.10.13.562235" class="uri">https://doi.org/10.1101/2023.10.13.562235</a>’.</p>
<div class="cell">
<div id="tbl-EAs" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-EAs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;9.1: Some commonly used enrichment analysis methods
</figcaption><div aria-describedby="tbl-EAs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<table class="do-not-create-environment cell table table-sm table-striped small" data-quarto-postprocess="true">
<thead><tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Category</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Method</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Tools</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">Notes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ORA</td>
<td style="text-align: left;">Hypergeometric test / Fisher’s exact test</td>
<td style="text-align: left;">DAVID (website), clusterProfiler (R package)</td>
<td style="text-align: left;">The most common methods used in enrichment analysis. Selecting a list of genes is required.</td>
</tr>
<tr class="even">
<td style="text-align: left;">FCS</td>
<td style="text-align: left;">Gene set enrichment analysis (GSEA)</td>
<td style="text-align: left;">GSEA (website)</td>
<td style="text-align: left;">GSEA creatively uses gene ranking, rather than selecting a list of genes, to identify statistically significant and concordant differences across gene sets.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FCS</td>
<td style="text-align: left;">Generalized reporter score-based analysis (GRSA/RSA)</td>
<td style="text-align: left;">ReporterScore (R package developed in this study)</td>
<td style="text-align: left;">Find significant metabolites (first report), pathways, and taxonomy based on the p-values for multi-omics data.</td>
</tr>
<tr class="even">
<td style="text-align: left;">FCS</td>
<td style="text-align: left;">Significance Analysis of Function and Expression (SAFE)</td>
<td style="text-align: left;">safe (R package)</td>
<td style="text-align: left;">SAFE assesses the significance of gene categories by calculating both local and global statistics from gene expression data.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FCS</td>
<td style="text-align: left;">Gene Set Analysis (GSA)</td>
<td style="text-align: left;">GSA (R Package)</td>
<td style="text-align: left;">GSA was proposed as an improvement of GSEA, using the “maxmean” statistic instead of the weighted sign KS statistic.</td>
</tr>
<tr class="even">
<td style="text-align: left;">FCS</td>
<td style="text-align: left;">Pathway Analysis with Down-weighting of Overlapping Genes</td>
<td style="text-align: left;">PADOG (R package)</td>
<td style="text-align: left;">PADOGA assumes that genes associated with fewer pathways have more significant effects than genes associated with more pathways.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FCS</td>
<td style="text-align: left;">Gene Set Variation Analysis (GSVA)</td>
<td style="text-align: left;">GSVA (R package)</td>
<td style="text-align: left;">A nonparametric, unsupervised method that transforms gene expression data into gene set scores for downstream differential pathway activity analysis.</td>
</tr>
<tr class="even">
<td style="text-align: left;">PT</td>
<td style="text-align: left;">Topology-based pathway enrichment analysis (TPEA)</td>
<td style="text-align: left;">TPEA (R package)</td>
<td style="text-align: left;">Integrate topological properties and global upstream/downstream positions of genes in pathways.</td>
</tr>
</tbody>
</table>
</div>
</div>
</figure>
</div>
</div>
<section id="reporterscore" class="level2" data-number="9.1"><h2 data-number="9.1" class="anchored" data-anchor-id="reporterscore">
<span class="header-section-number">9.1</span> ReporterScore</h2>
<p>Inspired by the classic RSA, we developed the improved Generalized Reporter Score-based Analysis (GRSA) method, implemented in the R package ReporterScore, along with comprehensive visualization methods and pathway databases. GRSA is a threshold-free method that works well with all types of biomedical features, such as genes, chemical compounds, and microbial species. GRSA works in the mixed (classic RSA) and directed modes (enhanced RSA). The directed mode uses signs of the reporter score to distinguish up-regulated or down-regulated pathways, which is more intuitive. Importantly, the GRSA supports multi-group and longitudinal experimental designs, because of the included multi-group-compatible statistical methods. The ReporterScore package also supports custom hierarchical and relational databases (e.g., table containing the correspondence between pathways and genes), providing extra flexibility for advanced users.</p>
<p>The workflow of GRSA in the ReporterScore package is shown in Figure <a href="#fig-workflow" class="quarto-xref">Figure&nbsp;<span>9.1</span></a>.</p>
<div id="fig-workflow" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-workflow-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/1-workflow.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-workflow-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.1: Generalized Reporter Score-based Enrichment Analysis for Omics Data
</figcaption></figure>
</div>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Asa12138/ReporterScore">ReporterScore</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"reporter_score_res"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/plot_report.html">plot_report</a></span><span class="op">(</span><span class="va">reporter_score_res</span>, rs_threshold <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-reporter" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-reporter-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="09-enrichment_files/figure-html/fig-reporter-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-reporter-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.2: GRSA result
</figcaption></figure>
</div>
</div>
</div>
<p>For convenience, the <code>ReporterScore</code> package provides an interface to the above-mentioned enrichment methods:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gsa_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/KO_gsa.html">KO_gsa</a></span><span class="op">(</span><span class="va">reporter_score_res</span>, p.adjust.method <span class="op">=</span> <span class="st">"BH"</span>, perm <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## perm= 10 / 100 </span></span>
<span><span class="co">## perm= 20 / 100 </span></span>
<span><span class="co">## perm= 30 / 100 </span></span>
<span><span class="co">## perm= 40 / 100 </span></span>
<span><span class="co">## perm= 50 / 100 </span></span>
<span><span class="co">## perm= 60 / 100 </span></span>
<span><span class="co">## perm= 70 / 100 </span></span>
<span><span class="co">## perm= 80 / 100 </span></span>
<span><span class="co">## perm= 90 / 100 </span></span>
<span><span class="co">## perm= 100 / 100</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/plot_enrich_res.html">plot_enrich_res</a></span><span class="op">(</span><span class="va">gsa_res</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-gsa" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-gsa-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="09-enrichment_files/figure-html/fig-gsa-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-gsa-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.3: GSA result
</figcaption></figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">padog_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/KO_padog.html">KO_padog</a></span><span class="op">(</span><span class="va">reporter_score_res</span>, p.adjust.method <span class="op">=</span> <span class="st">"none"</span>, perm <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">## Starting with 316 gene sets!</span></span>
<span><span class="co">## Analyzing 193 gene sets with 3 or more genes!</span></span>
<span><span class="co">## # of permutations used: 100 </span></span>
<span><span class="co">## 10 / 100 </span></span>
<span><span class="co">## 20 / 100 </span></span>
<span><span class="co">## 30 / 100 </span></span>
<span><span class="co">## 40 / 100 </span></span>
<span><span class="co">## 50 / 100 </span></span>
<span><span class="co">## 60 / 100 </span></span>
<span><span class="co">## 70 / 100 </span></span>
<span><span class="co">## 80 / 100 </span></span>
<span><span class="co">## 90 / 100 </span></span>
<span><span class="co">## 100 / 100</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/plot_enrich_res.html">plot_enrich_res</a></span><span class="op">(</span><span class="va">padog_res</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-padog" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-padog-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="09-enrichment_files/figure-html/fig-padog-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-padog-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.4: PADOG result
</figcaption></figure>
</div>
</div>
</div>
</section><section id="kegg" class="level2" data-number="9.2"><h2 data-number="9.2" class="anchored" data-anchor-id="kegg">
<span class="header-section-number">9.2</span> KEGG</h2>
<p><a href="https://www.genome.jp/kegg/brite.html">KEGG BRITE</a> is a collection of hierarchical classification systems capturing functional hierarchies of various biological objects, especially those represented as KEGG objects.</p>
<p>We collected k00001 KEGG Orthology (KO) table so that you can summaries each levels abundance. Use <code>load_KO_htable</code> to get KO_htable and use <code>update_KO_htable</code> to update. Use <code>up_level_KO</code> can upgrade to specific level in one of “pathway”, “module”, “level1”, “level2”, “level3”, “module1”, “module2”, “module3”.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">KO_htable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/load_htable.html">load_KO_htable</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">KO_htable</span>, n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:300px; overflow-x: scroll; width:100%; ">
<table class="table table-sm table-striped small" data-quarto-postprocess="true">
<thead><tr class="header">
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">level1_name</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">level2_name</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">level3_id</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">level3_name</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">KO_id</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">KO_name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K00844</td>
<td style="text-align: left;">HK; hexokinase [EC:2.7.1.1]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K12407</td>
<td style="text-align: left;">GCK; glucokinase [EC:2.7.1.2]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K00845</td>
<td style="text-align: left;">glk; glucokinase [EC:2.7.1.2]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K25026</td>
<td style="text-align: left;">glk; glucokinase [EC:2.7.1.2]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K01810</td>
<td style="text-align: left;">GPI, pgi; glucose-6-phosphate isomerase [EC:5.3.1.9]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K06859</td>
<td style="text-align: left;">pgi1; glucose-6-phosphate isomerase, archaeal [EC:5.3.1.9]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K13810</td>
<td style="text-align: left;">tal-pgi; transaldolase / glucose-6-phosphate isomerase [EC:2.2.1.2 5.3.1.9]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K15916</td>
<td style="text-align: left;">pgi-pmi; glucose/mannose-6-phosphate isomerase [EC:5.3.1.9 5.3.1.8]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K24182</td>
<td style="text-align: left;">PFK9; 6-phosphofructokinase [EC:2.7.1.11]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K00850</td>
<td style="text-align: left;">pfkA, PFK; 6-phosphofructokinase 1 [EC:2.7.1.11]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K16370</td>
<td style="text-align: left;">pfkB; 6-phosphofructokinase 2 [EC:2.7.1.11]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K21071</td>
<td style="text-align: left;">pfk, pfp; ATP-dependent phosphofructokinase / diphosphate-dependent phosphofructokinase [EC:2.7.1.11 2.7.1.90]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K00918</td>
<td style="text-align: left;">pfkC; ADP-dependent phosphofructokinase/glucokinase [EC:2.7.1.146 2.7.1.147]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K00895</td>
<td style="text-align: left;">pfp, PFP; diphosphate-dependent phosphofructokinase [EC:2.7.1.90]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K03841</td>
<td style="text-align: left;">FBP, fbp; fructose-1,6-bisphosphatase I [EC:3.1.3.11]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K02446</td>
<td style="text-align: left;">glpX; fructose-1,6-bisphosphatase II [EC:3.1.3.11]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K11532</td>
<td style="text-align: left;">glpX-SEBP; fructose-1,6-bisphosphatase II / sedoheptulose-1,7-bisphosphatase [EC:3.1.3.11 3.1.3.37]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K01086</td>
<td style="text-align: left;">fbp-SEBP; fructose-1,6-bisphosphatase I / sedoheptulose-1,7-bisphosphatase [EC:3.1.3.11 3.1.3.37]</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K04041</td>
<td style="text-align: left;">fbp3; fructose-1,6-bisphosphatase III [EC:3.1.3.11]</td>
</tr>
<tr class="even">
<td style="text-align: left;">Metabolism</td>
<td style="text-align: left;">Carbohydrate metabolism</td>
<td style="text-align: left;">map00010</td>
<td style="text-align: left;">Glycolysis / Gluconeogenesis</td>
<td style="text-align: left;">K01623</td>
<td style="text-align: left;">ALDO; fructose-bisphosphate aldolase, class I [EC:4.1.2.13]</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/plot_htable.html">plot_htable</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"ko"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-kegg" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-kegg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="09-enrichment_files/figure-html/fig-kegg-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-kegg-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.5: KEGG KO annotation categories
</figcaption></figure>
</div>
</div>
</div>
</section><section id="args-vfs" class="level2" data-number="9.3"><h2 data-number="9.3" class="anchored" data-anchor-id="args-vfs">
<span class="header-section-number">9.3</span> ARGs &amp; VFs</h2>
<p>For convenience, I also included the CARD database from <a href="https://card.mcmaster.ca/download/0/broadstreet-v3.2.8.tar.bz2" class="uri">https://card.mcmaster.ca/download/0/broadstreet-v3.2.8.tar.bz2</a>.</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">CARDinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReporterScore/man/load_CARDinfo.html">load_CARDinfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">my_sankey</span><span class="op">(</span><span class="fu">count</span><span class="op">(</span><span class="va">CARDinfo</span><span class="op">$</span><span class="va">ARO_index</span>, <span class="va">`AMR Gene Family`</span>, <span class="va">`Resistance Mechanism`</span>, <span class="va">`Drug Class`</span><span class="op">)</span>, topN <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-card" class="cell-output-display quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-card-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sankeyNetwork html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-432331342fe9306e2de2" style="width:100%;height:384px;"></div>
<script type="application/json" data-for="htmlwidget-432331342fe9306e2de2">{"x":{"links":{"source":[0,1,2,3,4,5,5,5,5,5,5,6,6,7,7,7,7,7,7,8,8,9,10,11,11],"target":[7,7,7,7,7,6,7,8,9,10,11,12,13,12,14,15,16,17,13,12,13,13,13,16,13],"value":[232,916,475,201,204,294,1966,564,176,72,73,5,289,170,473,1230,512,204,1405,22,542,176,72,1,72]},"nodes":{"name":["CTX-M beta-lactamase","OXA beta-lactamase","PDC beta-lactamase","SHV beta-lactamase","TEM beta-lactamase","other_AMR Gene Family","antibiotic efflux ","antibiotic inactivation ","antibiotic target alteration ","antibiotic target protection ","antibiotic target replacement ","other_Resistance Mechanism ","aminoglycoside antibiotic  ","other_Drug Class  ","carbapenem;cephalosporin;monobactam  ","carbapenem;cephalosporin;penam  ","cephalosporin  ","cephalosporin;monobactam;penam;penem  "],"group":["CTX-M beta-lactamase","OXA beta-lactamase","PDC beta-lactamase","SHV beta-lactamase","TEM beta-lactamase","other_AMR Gene Family","antibiotic efflux ","antibiotic inactivation ","antibiotic target alteration ","antibiotic target protection ","antibiotic target replacement ","other_Resistance Mechanism ","aminoglycoside antibiotic  ","other_Drug Class  ","carbapenem;cephalosporin;monobactam  ","carbapenem;cephalosporin;penam  ","cephalosporin  ","cephalosporin;monobactam;penam;penem  "]},"options":{"NodeID":"name","NodeGroup":"name","LinkGroup":null,"colourScale":"d3.scaleOrdinal().range(d3.schemeCategory20)","fontSize":7,"fontFamily":null,"fontColor":null,"nodeWidth":10,"nodePadding":10,"nodeStrokeWidth":1,"nodeCornerRadius":0,"dragX":false,"dragY":false,"numberFormat":",.5g","orderByPath":false,"units":"TWh","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":32,"zoom":false,"linkType":"bezier","curvature":0.5,"highlightChildLinks":false,"doubleclickTogglesChildren":false,"showNodeValues":true,"align":"justify","xAxisDomain":["AMR Gene Family","Resistance Mechanism","Drug Class"],"title":null,"nodeLabelMargin":2,"linkColor":"#A0A0A0","linkOpacity":0.5,"linkGradient":false,"nodeShadow":false,"scaleNodeBreadthsByString":false,"xScalingFactor":1,"yOrderComparator":null}},"evals":[],"jsHooks":[]}</script>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-card-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9.6: Sankey plot of ARG categories
</figcaption></figure>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./08-network.html" class="pagination-link  aria-label=" analysis="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Network analysis</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./10-element-cycle.html" class="pagination-link" aria-label="<span class='chapter-number'>10</span>&nbsp; <span class='chapter-title'>Elemental cycling</span>">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Elemental cycling</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/asa12138/pctax_tutorial/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/asa12138/pctax_tutorial/blob/main/09-enrichment.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>