<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>MetaNet: Network Analysis for Omics Data - 2&nbsp; Construction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./03-manipulation.html" rel="next">
<link href="./01-introduction.html" rel="prev">
<link href="./MetaNet.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script>
  MathJax = {
    tex: {
      inlineMath: [['$', '$'], ['\\(', '\\)']]
    }
  };
</script><script id="MathJax-script" async="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script><script src="site_libs/kePrint-0.0.1/kePrint.js"></script><link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
<meta property="og:title" content="MetaNet: Network Analysis for Omics Data - 2&nbsp; Construction">
<meta property="og:description" content="">
<meta property="og:site_name" content="MetaNet: Network Analysis for Omics Data">
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./02-construction.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Construction</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">MetaNet: Network Analysis for Omics Data</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/Asa12138/MetaNet_q_book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-construction.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Construction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-topology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Topology</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-stability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Stability</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-multi-omics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Multi-omics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#normalization" id="toc-normalization" class="nav-link active" data-scroll-target="#normalization"><span class="header-section-number">2.1</span> Pre-processing</a></li>
  <li>
<a href="#pairwise-relationship" id="toc-pairwise-relationship" class="nav-link" data-scroll-target="#pairwise-relationship"><span class="header-section-number">2.2</span> Pairwise relationship</a>
  <ul class="collapse">
<li><a href="#correlation" id="toc-correlation" class="nav-link" data-scroll-target="#correlation">Correlation</a></li>
  <li><a href="#sparcc" id="toc-sparcc" class="nav-link" data-scroll-target="#sparcc">SparCC</a></li>
  <li><a href="#distance" id="toc-distance" class="nav-link" data-scroll-target="#distance">Distance</a></li>
  <li><a href="#others" id="toc-others" class="nav-link" data-scroll-target="#others">Others</a></li>
  </ul>
</li>
  <li>
<a href="#build-network" id="toc-build-network" class="nav-link" data-scroll-target="#build-network"><span class="header-section-number">2.3</span> Build network</a>
  <ul class="collapse">
<li><a href="#normally-build" id="toc-normally-build" class="nav-link" data-scroll-target="#normally-build">Normally build</a></li>
  <li><a href="#multi-tables" id="toc-multi-tables" class="nav-link" data-scroll-target="#multi-tables">Multi-tables</a></li>
  <li><a href="#edgelist" id="toc-edgelist" class="nav-link" data-scroll-target="#edgelist">Edgelist</a></li>
  <li><a href="#others-1" id="toc-others-1" class="nav-link" data-scroll-target="#others-1">Others</a></li>
  </ul>
</li>
  <li><a href="#rmt-optimization" id="toc-rmt-optimization" class="nav-link" data-scroll-target="#rmt-optimization"><span class="header-section-number">2.4</span> RMT optimization</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/Asa12138/MetaNet_q_book/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/Asa12138/MetaNet_q_book/blob/main/02-construction.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Construction</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="normalization" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="normalization">
<span class="header-section-number">2.1</span> Pre-processing</h2>
<p>The <code>trans()</code> function contains many normalization methods, suitable for pre-processing of different omics, some refer to <code><a href="https://rdrr.io/pkg/vegan/man/decostand.html">vegan::decostand()</a></code>.</p>
<div class="cell">
<div id="tbl-normalization" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-normalization-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;2.1: Normalization methods used in omics
</figcaption><div aria-describedby="tbl-normalization-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:400px; overflow-x: scroll; width:100%; ">
<table class="table do-not-create-environment cell table-sm table-striped small" data-quarto-postprocess="true">
<thead><tr class="header">
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">Method</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">cpm</td>
<td style="text-align: left;">Counts per million</td>
</tr>
<tr class="even">
<td style="text-align: left;">minmax</td>
<td style="text-align: left;">linear transfer to (min, max)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">acpm</td>
<td style="text-align: left;">Counts per million, then asinh transfer</td>
</tr>
<tr class="even">
<td style="text-align: left;">log1</td>
<td style="text-align: left;">$log(n+1)$ transformat</td>
</tr>
<tr class="odd">
<td style="text-align: left;">total</td>
<td style="text-align: left;">divide by total</td>
</tr>
<tr class="even">
<td style="text-align: left;">max</td>
<td style="text-align: left;">divide by maximum</td>
</tr>
<tr class="odd">
<td style="text-align: left;">frequency</td>
<td style="text-align: left;">divide by total and multiply by the number of non-zero items, so that the average of non-zero entries is one.</td>
</tr>
<tr class="even">
<td style="text-align: left;">normalize</td>
<td style="text-align: left;">make margin sum of squares equal to one</td>
</tr>
<tr class="odd">
<td style="text-align: left;">range</td>
<td style="text-align: left;">standardize values into range (0,1) (same to minmax(0,1)). If all values are constant, they will be transformed to 0.</td>
</tr>
<tr class="even">
<td style="text-align: left;">rank</td>
<td style="text-align: left;">rank replaces abundance values by their increasing ranks leaving zeros unchanged.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rrank</td>
<td style="text-align: left;">rrank is similar but uses relative ranks with maximum 1.</td>
</tr>
<tr class="even">
<td style="text-align: left;">pa</td>
<td style="text-align: left;">scale x to presence/absence scale (0/1).</td>
</tr>
<tr class="odd">
<td style="text-align: left;">standardize</td>
<td style="text-align: left;">scale x to zero mean and unit variance.</td>
</tr>
<tr class="even">
<td style="text-align: left;">hellinger</td>
<td style="text-align: left;">square root of method = "total"</td>
</tr>
<tr class="odd">
<td style="text-align: left;">log</td>
<td style="text-align: left;">logarithmic transformation as suggested by Anderson et al. (2006): $log_b(x)+1$ for x&gt;0, where b is the base of the logarithm; zeros are left as zeros.</td>
</tr>
<tr class="even">
<td style="text-align: left;">alr</td>
<td style="text-align: left;">Additive log ratio ('alr') transformation (Aitchison 1986) reduces data skewness and compositionality bias.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">clr</td>
<td style="text-align: left;">centered log ratio ('clr') transformation proposed by Aitchison (1986) reduces data skewness and compositionality bias.</td>
</tr>
<tr class="even">
<td style="text-align: left;">rclr</td>
<td style="text-align: left;">robust clr ('rclr') is similar to regular clr (see above) but allows data that contains zeroes.</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
</figure>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">trans</span><span class="op">(</span><span class="va">otutab</span>, method <span class="op">=</span> <span class="st">"log1"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:300px; overflow-x: scroll; width:100%; ">
<table class="table table-sm table-striped small" data-quarto-postprocess="true">
<thead><tr class="header">
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;"></th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">NS1</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">NS2</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">NS3</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">NS4</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">NS5</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">NS6</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">WS1</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">WS2</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">WS3</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">WS4</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">WS5</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">WS6</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">CS1</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">CS2</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">CS3</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">CS4</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">CS5</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">CS6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">s__un_f__Thermomonosporaceae</td>
<td style="text-align: right;">6.996682</td>
<td style="text-align: right;">7.560601</td>
<td style="text-align: right;">6.698268</td>
<td style="text-align: right;">7.211557</td>
<td style="text-align: right;">6.970730</td>
<td style="text-align: right;">6.976348</td>
<td style="text-align: right;">7.133296</td>
<td style="text-align: right;">7.376508</td>
<td style="text-align: right;">7.193686</td>
<td style="text-align: right;">6.848005</td>
<td style="text-align: right;">7.118016</td>
<td style="text-align: right;">6.919684</td>
<td style="text-align: right;">7.746733</td>
<td style="text-align: right;">7.831617</td>
<td style="text-align: right;">7.444249</td>
<td style="text-align: right;">7.588830</td>
<td style="text-align: right;">7.266827</td>
<td style="text-align: right;">7.331715</td>
</tr>
<tr class="even">
<td style="text-align: left;">s__Pelomonas_puraquae</td>
<td style="text-align: right;">7.582229</td>
<td style="text-align: right;">7.118826</td>
<td style="text-align: right;">7.767687</td>
<td style="text-align: right;">7.712891</td>
<td style="text-align: right;">7.973844</td>
<td style="text-align: right;">7.512071</td>
<td style="text-align: right;">6.469250</td>
<td style="text-align: right;">6.206576</td>
<td style="text-align: right;">7.115582</td>
<td style="text-align: right;">7.158514</td>
<td style="text-align: right;">6.860664</td>
<td style="text-align: right;">6.455199</td>
<td style="text-align: right;">7.174724</td>
<td style="text-align: right;">7.324490</td>
<td style="text-align: right;">6.739337</td>
<td style="text-align: right;">7.029088</td>
<td style="text-align: right;">7.302496</td>
<td style="text-align: right;">7.069023</td>
</tr>
<tr class="odd">
<td style="text-align: left;">s__Rhizobacter_bergeniae</td>
<td style="text-align: right;">6.378426</td>
<td style="text-align: right;">6.129050</td>
<td style="text-align: right;">6.791221</td>
<td style="text-align: right;">6.804614</td>
<td style="text-align: right;">7.112327</td>
<td style="text-align: right;">6.749931</td>
<td style="text-align: right;">6.405228</td>
<td style="text-align: right;">6.154858</td>
<td style="text-align: right;">6.976348</td>
<td style="text-align: right;">6.936343</td>
<td style="text-align: right;">6.741701</td>
<td style="text-align: right;">6.508769</td>
<td style="text-align: right;">6.937314</td>
<td style="text-align: right;">7.497207</td>
<td style="text-align: right;">6.910751</td>
<td style="text-align: right;">7.090910</td>
<td style="text-align: right;">7.085902</td>
<td style="text-align: right;">6.637258</td>
</tr>
<tr class="even">
<td style="text-align: left;">s__Flavobacterium_terrae</td>
<td style="text-align: right;">5.501258</td>
<td style="text-align: right;">5.459586</td>
<td style="text-align: right;">7.501634</td>
<td style="text-align: right;">6.513230</td>
<td style="text-align: right;">7.276556</td>
<td style="text-align: right;">6.198479</td>
<td style="text-align: right;">5.765191</td>
<td style="text-align: right;">7.563720</td>
<td style="text-align: right;">7.309212</td>
<td style="text-align: right;">6.903747</td>
<td style="text-align: right;">6.359574</td>
<td style="text-align: right;">5.886104</td>
<td style="text-align: right;">6.985642</td>
<td style="text-align: right;">7.105786</td>
<td style="text-align: right;">6.626718</td>
<td style="text-align: right;">6.049734</td>
<td style="text-align: right;">6.940222</td>
<td style="text-align: right;">7.253470</td>
</tr>
<tr class="odd">
<td style="text-align: left;">s__un_g__Rhizobacter</td>
<td style="text-align: right;">7.267525</td>
<td style="text-align: right;">6.023448</td>
<td style="text-align: right;">6.280396</td>
<td style="text-align: right;">6.633318</td>
<td style="text-align: right;">7.162397</td>
<td style="text-align: right;">6.228511</td>
<td style="text-align: right;">6.222576</td>
<td style="text-align: right;">6.381816</td>
<td style="text-align: right;">6.100319</td>
<td style="text-align: right;">6.431331</td>
<td style="text-align: right;">6.489205</td>
<td style="text-align: right;">6.063785</td>
<td style="text-align: right;">7.032624</td>
<td style="text-align: right;">7.277939</td>
<td style="text-align: right;">6.311735</td>
<td style="text-align: right;">6.369901</td>
<td style="text-align: right;">7.008505</td>
<td style="text-align: right;">6.806829</td>
</tr>
<tr class="even">
<td style="text-align: left;">s__un_o__Burkholderiales</td>
<td style="text-align: right;">6.787845</td>
<td style="text-align: right;">6.527958</td>
<td style="text-align: right;">6.715383</td>
<td style="text-align: right;">6.816736</td>
<td style="text-align: right;">7.315218</td>
<td style="text-align: right;">6.937314</td>
<td style="text-align: right;">5.463832</td>
<td style="text-align: right;">5.533390</td>
<td style="text-align: right;">5.886104</td>
<td style="text-align: right;">5.945421</td>
<td style="text-align: right;">5.961005</td>
<td style="text-align: right;">5.863631</td>
<td style="text-align: right;">6.313548</td>
<td style="text-align: right;">6.293419</td>
<td style="text-align: right;">6.169611</td>
<td style="text-align: right;">6.327937</td>
<td style="text-align: right;">6.242223</td>
<td style="text-align: right;">6.208590</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
<p><code>rm_low()</code>, <code>guolv()</code> and <code>hebing()</code> functions can help filter or aggregate the omics data.</p>
</section><section id="pairwise-relationship" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="pairwise-relationship">
<span class="header-section-number">2.2</span> Pairwise relationship</h2>
<p>Pairwise relationship is the basis of network construction.</p>
<p>How to determine the pairwise relationship, because the experimental data is generally relatively rare, we mainly relying on statistical inference.</p>
<p>At present, there are mainly two statistical ways:</p>
<ul>
<li><p>Similarity or correlation <span class="citation" data-cites="faustMicrobialInteractionsNetworks2012">(<a href="references.html#ref-faustMicrobialInteractionsNetworks2012" role="doc-biblioref"><em>1</em></a>)</span>. For example: Spearman, Pearson, Bray-Curtis, etc. The similarity matrix between paired features can be calculated, and the randomized data can be used to repeatedly calculate the significance. Finally, meaningful similarities are retained in network.</p></li>
<li><p>Regression. Divide features into source and target, and use multiple regression to calculate the relationship between features.</p></li>
</ul>
<p>Some tools use special methods to optimize the network construction, such as <strong>SparCC</strong>, etc.</p>
<section id="correlation" class="level3"><h3 class="anchored" data-anchor-id="correlation">Correlation</h3>
<p><strong>Correlation</strong> is a statistical term describing the degree to which two variables move in coordination with one-another.</p>
<p>Correlation calculation is the first step in all omics network analysis software, there are many method to get <span class="math inline">\(\rho\)</span> and <span class="math inline">\(p\)</span>-value. However, as the data size of omics grow larger and larger, some methods became very time and computational resource consuming.</p>
<p>Here, we provide the <code>c_net_calculate()</code> function for one single table or two tables to calculate correlation <strong>fastly</strong> (fig-2-packages-comparison), which will return a <code>coor</code> object including <span class="math inline">\(\rho\)</span> and <span class="math inline">\(p\)</span>-value.</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># single table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">otutab</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">totu</span></span>
<span><span class="fu">c_net_calculate</span><span class="op">(</span><span class="va">totu</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">corr</span></span>
<span><span class="va">corr</span></span>
<span><span class="co">## Correlation table:</span></span>
<span><span class="co">## Table dimensions: 485 rows, 485 columns</span></span>
<span><span class="co">## Use method:  spearman</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># two tables</span></span>
<span><span class="va">metadata</span><span class="op">[</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op">-&gt;</span> <span class="va">env</span></span>
<span><span class="fu">c_net_calculate</span><span class="op">(</span><span class="va">totu</span>, <span class="va">env</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">corr2</span></span>
<span><span class="va">corr2</span></span>
<span><span class="co">## Correlation table:</span></span>
<span><span class="co">## Table dimensions: 485 rows, 8 columns</span></span>
<span><span class="co">## Use method:  spearman</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If your input table is too large, the calculation may take a longer time, you can save the result and read it later.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_calculate</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">1000</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">corr</span></span>
<span></span>
<span><span class="co"># save the result</span></span>
<span><span class="fu">save_corr</span><span class="op">(</span><span class="va">corr</span>, <span class="st">"../saved_corr1"</span><span class="op">)</span></span>
<span><span class="fu">read_corr</span><span class="op">(</span><span class="st">"../saved_corr1"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">corr_new</span></span>
<span></span>
<span><span class="co"># they are identical</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span><span class="op">(</span><span class="va">corr</span>, <span class="va">corr_new</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>According to the test, a table with 1,000 features will cost about 0.2s and to calculate the correlation, and the size of saved .corr file is about 6 MB.</p>
<p>Time, memory consumption, and saved file size approximately increase linearly with the square of the number of features, as shown in the following figure. By the way, these consumption increase a little bit with the number of samples.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">feature_number</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">500</span>, <span class="fl">1000</span>, <span class="fl">2000</span>, <span class="fl">3000</span>, <span class="fl">4000</span><span class="op">)</span></span>
<span><span class="va">time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">memory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">save_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">feature_number</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">i</span> <span class="op">*</span> <span class="fl">50</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="va">i</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">tmp</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="fu">c_net_calculate</span><span class="op">(</span><span class="va">tmp</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">tmp_corr</span></span>
<span>  <span class="op">}</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">time1</span></span>
<span>  <span class="va">time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">time</span>, <span class="va">time1</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">memory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">memory</span>, <span class="fu"><a href="https://rdrr.io/r/utils/object.size.html">object.size</a></span><span class="op">(</span><span class="va">tmp_corr</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu">save_corr</span><span class="op">(</span><span class="va">tmp_corr</span>, <span class="st">"../tmp_corr.corr"</span><span class="op">)</span></span>
<span>  <span class="va">save_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">save_size</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.info.html">file.size</a></span><span class="op">(</span><span class="st">"../tmp_corr.corr"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  feature_number <span class="op">=</span> <span class="va">feature_number</span>,</span>
<span>  `time(s)` <span class="op">=</span> <span class="va">time</span>, `memory(MB)` <span class="op">=</span> <span class="va">memory</span> <span class="op">/</span> <span class="fl">1024</span> <span class="op">/</span> <span class="fl">1024</span>,</span>
<span>  `save_size(MB)` <span class="op">=</span> <span class="va">save_size</span> <span class="op">/</span> <span class="fl">1024</span> <span class="op">/</span> <span class="fl">1024</span>, check.names <span class="op">=</span> <span class="cn">F</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="fu">melt</span><span class="op">(</span><span class="va">plot_data</span>, id.vars <span class="op">=</span> <span class="st">"feature_number"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">plot_data</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">feature_number</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_line</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">variable</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Time and memory consumption of `c_net_calculate`"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"feature number"</span>, y <span class="op">=</span> <span class="st">"value"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-2-time_memory" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-time_memory-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/2-1.time_memory.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-time_memory-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.1: Time and memory consumption of <code>c_net_calculate</code>
</figcaption></figure>
</div>
</div>
</div>
<p>The calculation speed of <code>c_net_calculate</code> is much faster than other R packages, especially when the number of features is large:</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-2-packages-comparison" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-packages-comparison-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-packages-comparison-1.png" class="img-fluid figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-packages-comparison-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.2: Comparison of correlation calculation speed between different packages
</figcaption></figure>
</div>
</div>
</div>
</section><section id="sparcc" class="level3"><h3 class="anchored" data-anchor-id="sparcc">SparCC</h3>
<p>SparCC fits the Dirichlet distribution to the observed data, and iteratively calculates the proportion and correlation of species several times. The resulting correlation is the median of the distribution. <span class="math inline">\(p\)</span>-values were calculated using the bootstrap method.</p>
<p>This metric is said to be more useful with non-normal microbiome data.</p>
<p><span class="math inline">\(D(x_i,x_j)=var(\log(\frac{x_i}{x_j}))\)</span></p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_calculate</span><span class="op">(</span><span class="va">totu</span>, method <span class="op">=</span> <span class="st">"sparcc"</span><span class="op">)</span></span>
<span><span class="co">## NULL corr object.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="distance" class="level3"><h3 class="anchored" data-anchor-id="distance">Distance</h3>
<p>We should understand the dissimilarity indices (distance) for community. The <code><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegan::vegdist()</a></code> function provides a wide range of distance metrics based on abundance calculations (<a href="#tbl-dissimilarity" class="quarto-xref">Table&nbsp;<span>2.2</span></a>). Additionally, there are distance metrics that incorporate phylogenetic data such as “unifrac,” “beta_mpd,” “beta_mntd,” and “phylosor”, etc.</p>
<p>And the similarity indices are the inverse of the distance indices, often as <span class="math inline">\(1-distance\)</span>. We can also use <code>c_net_calculate()</code> to calculate the distance. The the similarity indices are more suitable for relationship between samples.</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_calculate</span><span class="op">(</span><span class="va">otutab</span>, method <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">sim_df</span></span>
<span><span class="va">sim_df</span></span>
<span><span class="co">## Similarity (1-Distance) table:</span></span>
<span><span class="co">## Table dimensions: 18 rows, 18 columns</span></span>
<span><span class="co">## Use method:  bray</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div id="tbl-dissimilarity" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-dissimilarity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;2.2: Some dissimilarity indices for community
</figcaption><div aria-describedby="tbl-dissimilarity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:500px; overflow-x: scroll; width:100%; ">
<table class="table do-not-create-environment cell table-sm table-striped small" data-quarto-postprocess="true">
<thead><tr class="header">
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">Distance.Coefficient</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Manhattan</td>
<td style="text-align: left;">Manhattan distance calculates the sum of absolute differences between coordinates in each dimension. It is suitable for data with categorical variables or attributes.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Euclidean</td>
<td style="text-align: left;">Euclidean distance calculates the straight-line distance between two points in a multidimensional space. It is commonly used when data are continuous and have no categorical attributes.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Canberra</td>
<td style="text-align: left;">Canberra distance is a weighted version of Manhattan distance that emphasizes the importance of small differences between coordinates. It is often used for ecological data analysis.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Clark</td>
<td style="text-align: left;">Clark distance measures the proportion of non-zero attributes that are different between two samples.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Bray</td>
<td style="text-align: left;">Bray distance calculates the dissimilarity between two samples based on the sum of absolute differences in abundances.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Kulczynski</td>
<td style="text-align: left;">Kulczynski distance measures the similarity between two samples based on the arithmetic mean of proportions of common attributes.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Jaccard</td>
<td style="text-align: left;">Jaccard distance measures dissimilarity between two samples based on the proportion of unique attributes. It is commonly used in ecology for binary data.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Gower</td>
<td style="text-align: left;">Gower distance is a generalization of Manhattan distance for mixed data types, including categorical and continuous variables.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">AltGower</td>
<td style="text-align: left;">AltGower distance is an alternative form of Gower distance that uses an alternative method to standardize continuous variables.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Morisita</td>
<td style="text-align: left;">Morisita distance measures dissimilarity between two samples based on the frequency of shared attributes, with emphasis on rare attributes.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Horn</td>
<td style="text-align: left;">Horn distance measures dissimilarity between two samples based on the geometric mean of proportions of shared attributes.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Mountford</td>
<td style="text-align: left;">Mountford distance measures dissimilarity between two samples based on the reciprocal of the arithmetic mean of proportions of shared attributes.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Raup</td>
<td style="text-align: left;">Raup distance measures dissimilarity between two samples based on the probability of shared attributes.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Binomial</td>
<td style="text-align: left;">Binomial distance measures dissimilarity between two samples based on the probability of observing shared attributes.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Chao</td>
<td style="text-align: left;">Chao distance is a modification of Jaccard distance that adjusts for undersampling of rare species.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Cao</td>
<td style="text-align: left;">Cao distance is a modification of Chao distance that incorporates species abundance information.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mahalanobis</td>
<td style="text-align: left;">Mahalanobis distance measures the distance between two samples in multidimensional space, accounting for correlation between variables.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chisq</td>
<td style="text-align: left;">Chisq distance calculates the dissimilarity between two samples based on the chi-squared distance between categorical variables.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Chord</td>
<td style="text-align: left;">Chord distance calculates dissimilarity between two samples based on the angles between points in multidimensional space.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Hellinger</td>
<td style="text-align: left;">Hellinger distance measures dissimilarity between two samples based on the square root of the sum of squared differences in square root-transformed abundances.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aitchison</td>
<td style="text-align: left;">Aitchison distance is a transformation of compositional data that allows for Euclidean distance calculation in log-ratio space.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Robust Aitchison</td>
<td style="text-align: left;">Robust Aitchison distance is a robust version of Aitchison distance that reduces the influence of outliers in the data.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Unifrac</td>
<td style="text-align: left;">Unifrac distance measures dissimilarity between microbial communities based on evolutionary distances in a phylogenetic tree.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Beta MPD</td>
<td style="text-align: left;">Beta MPD (Mean Phylogenetic Distance) measures the phylogenetic diversity between two communities based on the mean phylogenetic distance of species pairs.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Beta MNTD</td>
<td style="text-align: left;">Beta MNTD (Mean Nearest Taxon Distance) measures the phylogenetic turnover between two communities based on the mean nearest taxon distance.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Phylosor</td>
<td style="text-align: left;">Phylosor distance measures the dissimilarity between communities based on the proportion of shared branches in a phylogenetic tree.</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
</figure>
</div>
</div>
</section><section id="others" class="level3"><h3 class="anchored" data-anchor-id="others">Others</h3>
<p>There are some other methods available for network construction in <a href="https://github.com/stefpeschel/NetCoMi"><strong>NetCoMi</strong></a>:</p>
<ul>
<li><p>Biweight Midcorrelation <a href="https://rdrr.io/cran/WGCNA/man/bicor.html"><code>bicor()</code></a> from <code>WGCNA</code> package</p></li>
<li><p>CCLasso (<a href="https://github.com/huayingfang/CCLasso">R code on GitHub</a>)</p></li>
<li><p>CCREPE (<a href="https://bioconductor.org/packages/release/bioc/html/ccrepe.html"><code>ccrepe</code></a> package)</p></li>
<li><p>SPRING (<a href="https://github.com/GraceYoon/SPRING"><code>SPRING</code></a> package)</p></li>
<li><p>gCoda (<a href="https://github.com/huayingfang/gCoda">R code on GitHub</a>)</p></li>
<li><p>propr (<a href="https://cran.r-project.org/web/packages/propr/index.html"><code>propr</code></a> package)</p></li>
<li><p>Kullback-Leibler divergence (KLD) (<a href="https://rdrr.io/cran/LaplacesDemon/man/KLD.html"><code>KLD()</code></a> from <code>LaplacesDemon</code> package)</p></li>
<li><p>Jeffrey divergence (own code using <a href="https://rdrr.io/cran/LaplacesDemon/man/KLD.html"><code>KLD()</code></a> from <code>LaplacesDemon</code> package)</p></li>
<li><p>Jensen-Shannon divergence (own code using <a href="https://rdrr.io/cran/LaplacesDemon/man/KLD.html"><code>KLD()</code></a> from <code>LaplacesDemon</code> package)</p></li>
</ul></section></section><section id="build-network" class="level2" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="build-network">
<span class="header-section-number">2.3</span> Build network</h2>
<section id="normally-build" class="level3"><h3 class="anchored" data-anchor-id="normally-build">Normally build</h3>
<p>If you have done the <code>c_net_calculate()</code> step, then you can get a network (<code>metanet</code> object, inherit from <code>igraph</code> object) easily by <code>c_net_build()</code>. Some common attributes will be set automatically.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_build</span><span class="op">(</span><span class="va">corr</span>, r_threshold <span class="op">=</span> <span class="fl">0.6</span>, p_threshold <span class="op">=</span> <span class="fl">0.05</span>, delete_single <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net</span></span>
<span><span class="va">co_net</span></span>
<span><span class="co">## =================================== metanet ==================================== </span></span>
<span><span class="co">## IGRAPH 83c9230 UNW- 483 1513 -- </span></span>
<span><span class="co">## + attr: n_type (g/c), name (v/c), v_group (v/c), v_class (v/c), size</span></span>
<span><span class="co">## | (v/n), label (v/c), shape (v/c), color (v/c), id (e/n), from (e/c),</span></span>
<span><span class="co">## | to (e/c), weight (e/n), cor (e/n), p.value (e/n), e_type (e/c), width</span></span>
<span><span class="co">## | (e/n), color (e/c), e_class (e/c), lty (e/n)</span></span>
<span><span class="co">## + edges from 83c9230 (vertex names):</span></span>
<span><span class="co">## [1] s__un_f__Thermomonosporaceae--s__Actinocorallia_herbida    </span></span>
<span><span class="co">## [2] s__un_f__Thermomonosporaceae--s__Kribbella_catacumbae      </span></span>
<span><span class="co">## [3] s__un_f__Thermomonosporaceae--s__Kineosporia_rhamnosa      </span></span>
<span><span class="co">## [4] s__un_f__Thermomonosporaceae--s__un_f__Micromonosporaceae  </span></span>
<span><span class="co">## [5] s__un_f__Thermomonosporaceae--s__Flavobacterium_saliperosum</span></span>
<span><span class="co">## + ... omitted several edges</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">co_net</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-2-conet1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-conet1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-conet1-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-conet1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.3: Simple co-occurrence network
</figcaption></figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_build</span><span class="op">(</span><span class="va">corr2</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">co_net2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-2-conet2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-conet2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-conet2-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-conet2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.4: Simple bipartite network
</figcaption></figure>
</div>
</div>
</div>
</section><section id="multi-tables" class="level3"><h3 class="anchored" data-anchor-id="multi-tables">Multi-tables</h3>
<p>When you have more than two tables for correlation network analysis, you can choose the <code>multi_net_build()</code> to calculate and build network.</p>
<p>For subsequent <strong>multi-omics analysis</strong>, see <a href="07-multi-omics.html" class="quarto-xref"><span>Chapter 7</span></a>.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"multi_test"</span>, package <span class="op">=</span> <span class="st">"MetaNet"</span><span class="op">)</span></span>
<span><span class="co"># microbiome</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">micro</span><span class="op">)</span></span>
<span><span class="co">## [1] 18 50</span></span>
<span><span class="co"># metabolome</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">metab</span><span class="op">)</span></span>
<span><span class="co">## [1] 18 50</span></span>
<span><span class="co"># transcriptome</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">transc</span><span class="op">)</span></span>
<span><span class="co">## [1] 18 50</span></span>
<span></span>
<span><span class="fu">multi_net_build</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  Microbiome <span class="op">=</span> <span class="va">micro</span>,</span>
<span>  Metabolome <span class="op">=</span> <span class="va">metab</span>,</span>
<span>  Transcriptome <span class="op">=</span> <span class="va">transc</span></span>
<span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">multi1</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">multi1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-2-multinet1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-multinet1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-multinet1-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-multinet1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.5: Simple multi-omics network
</figcaption></figure>
</div>
</div>
</div>
</section><section id="edgelist" class="level3"><h3 class="anchored" data-anchor-id="edgelist">Edgelist</h3>
<p>If you already get the pairwise relationship of data from other approaches (database), you can form it into a edgelist and use <code>c_net_from_edgelist</code> to build network. It is useful for following analysis.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"edgelist"</span>, package <span class="op">=</span> <span class="st">"MetaNet"</span><span class="op">)</span></span>
<span><span class="va">dnet</span> <span class="op">&lt;-</span> <span class="fu">c_net_from_edgelist</span><span class="op">(</span><span class="va">arc_count</span>, direct <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">dnet</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-2-directnet" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-directnet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-directnet-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-directnet-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.6: Simple directed network
</figcaption></figure>
</div>
</div>
</div>
</section><section id="others-1" class="level3"><h3 class="anchored" data-anchor-id="others-1">Others</h3>
<p>If you have network in other format, such as pajek, lgl, graphml, dimacs, graphdb, gml and dl, you can use <code>read.graph</code> to read it as <code>igraph</code> object. And you can also use <code>graph_from_*</code> functions to get <code>igraph</code> object, then use <code>c_net_update</code> to transform it into <code>metanet</code> object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fu">graph_from_lcf</span><span class="op">(</span><span class="fl">12</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="op">-</span><span class="fl">5</span><span class="op">)</span>, <span class="fl">6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">g1</span><span class="op">)</span></span>
<span><span class="co">## [1] "igraph"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">g1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">c_net_update</span><span class="op">(</span><span class="va">g1</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">g2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span></span>
<span><span class="co">## [1] "metanet" "igraph"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">g2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-2-graph" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-graph-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-2-graph" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-2-graph-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-2-graph-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-graph-1.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-2-graph">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-2-graph-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) igraph plot
</figcaption></figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-2-graph" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-2-graph-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-2-graph-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="02-construction_files/figure-html/fig-2-graph-2.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-2-graph">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-2-graph-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) metanet plot
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-graph-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.7: igraph plot and metanet plot
</figcaption></figure>
</div>
</section></section><section id="rmt-optimization" class="level2" data-number="2.4"><h2 data-number="2.4" class="anchored" data-anchor-id="rmt-optimization">
<span class="header-section-number">2.4</span> RMT optimization</h2>
<p>The correlation-based relevance network method is most commonly used largely due to its simple calculation procedure and noise tolerance. However, most studies involving relevance network analysis use arbitrary thresholds (usually, we use r&gt;0.6, p&lt;0.05), and thus the constructed networks are subjective rather than objective.</p>
<p>This problem has been partially solved by a random matrix theory (RMT)-based approach (<a href="#fig-2-RMT" class="quarto-xref">Figure&nbsp;<span>2.8</span></a>), which is able to automatically identify a proper threshold for network construction from micro-array data (also other omics data) <span class="citation" data-cites="dengMolecularEcologicalNetwork2012">(<a href="references.html#ref-dengMolecularEcologicalNetwork2012" role="doc-biblioref"><em>2</em></a>)</span>.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-2-RMT" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-RMT-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/2-1.RMT.png" style="width:100.0%;height:80.0%" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-RMT-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.8: Random matrix theory (RMT)-based approach
</figcaption></figure>
</div>
</div>
</div>
<p>use <code>RMT_threshold()</code> , we can find a proper <code>r_threshold</code> to build the network with smallest noise.</p>
<p>the bigger <code>log_LE</code>, less <code>log_LW</code>, less <code>log_see</code>, bigger <code>p_ks_test</code> indicate the better <code>r_threshold</code> for a meaningful network construction.You can change the threshold range to calculate more finely.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">RMT_threshold</span><span class="op">(</span><span class="va">corr</span>,</span>
<span>  out_dir <span class="op">=</span> <span class="st">"../RMT_temp/"</span>,</span>
<span>  min_threshold <span class="op">=</span> <span class="fl">0.5</span>, max_threshold <span class="op">=</span> <span class="fl">0.9</span>, step <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">rmt_res</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">rmt_res</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-2-RMT-res" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-RMT-res-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/2-2.RMT.png" class="img-fluid figure-img" style="width:80.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-RMT-res-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.9: RMT_threshold result from 0.5 to 0.9
</figcaption></figure>
</div>
</div>
</div>
<p>You can set the <code>gif=TRUE</code> in <code>RMT_threshold()</code> and get a gif file to observe the distribution of eigenvalue spacing for different r-thresholds.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-2-RMT-res2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-2-RMT-res2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/2-3.rmt_nnsd.gif" class="img-fluid figure-img" style="width:70.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-2-RMT-res2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.10: the distribution of eigenvalue spacing from 0.5 to 0.9
</figcaption></figure>
</div>
</div>
</div>


<div id="refs" class="references csl-bib-body" role="list" style="display: none">
<div id="ref-faustMicrobialInteractionsNetworks2012" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">K. Faust, J. Raes, Microbial interactions: From networks to models. <em>Nature Reviews Microbiology</em> (2012), doi:<a href="https://doi.org/10.1038/nrmicro2832">10.1038/nrmicro2832</a>.</div>
</div>
<div id="ref-dengMolecularEcologicalNetwork2012" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div>
<div class="csl-right-inline">Y. Deng, Y. Jiang, Y. Yang, Z. He, <em>et al.</em>, Molecular ecological network analyses. <em>BMC bioinformatics</em> (2012), doi:<a href="https://doi.org/10.1186/1471-2105-13-113">10.1186/1471-2105-13-113</a>.</div>
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./01-introduction.html" class="pagination-link  aria-label=">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./03-manipulation.html" class="pagination-link" aria-label="<span class='chapter-number'>3</span>&nbsp; <span class='chapter-title'>Manipulation</span>">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulation</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/Asa12138/MetaNet_q_book/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/Asa12138/MetaNet_q_book/blob/main/02-construction.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>