<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>MetaNet: Network Analysis for Omics Data - 3&nbsp; Manipulation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./04-visualization.html" rel="next">
<link href="./02-construction.html" rel="prev">
<link href="./MetaNet.svg" rel="icon" type="image/svg+xml">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script>
  MathJax = {
    tex: {
      inlineMath: [['$', '$'], ['\\(', '\\)']]
    }
  };
</script><script id="MathJax-script" async="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js">
</script><script src="site_libs/kePrint-0.0.1/kePrint.js"></script><link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">
<script src="site_libs/htmlwidgets-1.6.1/htmlwidgets.js"></script><script src="site_libs/sankeyNetwork-binding-0.3.2/sankeyNetwork.js"></script><script src="site_libs/d3-4.13.0/d3.min.js"></script><script src="site_libs/sankey-0.1/sankey.js"></script>
<meta property="og:title" content="MetaNet: Network Analysis for Omics Data - 3&nbsp; Manipulation">
<meta property="og:description" content="">
<meta property="og:site_name" content="MetaNet: Network Analysis for Omics Data">
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./03-manipulation.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulation</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">MetaNet: Network Analysis for Omics Data</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/Asa12138/MetaNet_q_book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02-construction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Construction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03-manipulation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05-topology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Topology</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06-stability.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Stability</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07-multi-omics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Multi-omics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#attributes" id="toc-attributes" class="nav-link active" data-scroll-target="#attributes"><span class="header-section-number">3.1</span> Attributes</a></li>
  <li><a href="#annotation" id="toc-annotation" class="nav-link" data-scroll-target="#annotation"><span class="header-section-number">3.2</span> Annotation</a></li>
  <li><a href="#set-attributes" id="toc-set-attributes" class="nav-link" data-scroll-target="#set-attributes"><span class="header-section-number">3.3</span> Set attributes</a></li>
  <li><a href="#filter-sub-net" id="toc-filter-sub-net" class="nav-link" data-scroll-target="#filter-sub-net"><span class="header-section-number">3.4</span> Filter (Sub-net)</a></li>
  <li><a href="#union-combine" id="toc-union-combine" class="nav-link" data-scroll-target="#union-combine"><span class="header-section-number">3.5</span> Union (Combine)</a></li>
  <li><a href="#skeleton" id="toc-skeleton" class="nav-link" data-scroll-target="#skeleton"><span class="header-section-number">3.6</span> Skeleton</a></li>
  <li><a href="#save" id="toc-save" class="nav-link" data-scroll-target="#save"><span class="header-section-number">3.7</span> Save</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/Asa12138/MetaNet_q_book/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/Asa12138/MetaNet_q_book/blob/main/03-manipulation.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Manipulation</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p>After we build a network, we can see the class of our network is <code>metanet</code>, which is a derived object from <code>igraph</code>.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">otutab</span>, package <span class="op">=</span> <span class="st">"pcutils"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">otutab</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">totu</span></span>
<span><span class="fu">c_net_calculate</span><span class="op">(</span><span class="va">totu</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">corr</span></span>
<span><span class="fu">c_net_build</span><span class="op">(</span><span class="va">corr</span>, r_threshold <span class="op">=</span> <span class="fl">0.6</span>, p_threshold <span class="op">=</span> <span class="fl">0.05</span>, delete_single <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">co_net</span><span class="op">)</span></span>
<span><span class="co">## [1] "metanet" "igraph"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Functions used for <code>igraph</code> can also use on the <code>metanet</code> object, refer to the <a href="https://r.igraph.org/">igraph manual</a> for more details. Besides, there are some functions in <code>MetaNet</code> for network manipulation, such as setting attributes, filtering, summarizing, and exporting:</p>
<section id="attributes" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="attributes">
<span class="header-section-number">3.1</span> Attributes</h2>
<p>We could get the attributes of whole network, each vertex and each edge by <code>get_*()</code> as data.frame:</p>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># get network attributes</span></span>
<span><span class="fu">get_n</span><span class="op">(</span><span class="va">co_net</span><span class="op">)</span></span>
<span><span class="co">##   n_type</span></span>
<span><span class="co">## 1 single</span></span>
<span><span class="co"># get vertex attributes</span></span>
<span><span class="fu">get_v</span><span class="op">(</span><span class="va">co_net</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##                           name  v_group  v_class size</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae v_group1 v_class1    4</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae v_group1 v_class1    4</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae v_group1 v_class1    4</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae v_group1 v_class1    4</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter v_group1 v_class1    4</span></span>
<span><span class="co">##                          label  shape   color</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae circle #a6bce3</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae circle #a6bce3</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae circle #a6bce3</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae circle #a6bce3</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter circle #a6bce3</span></span>
<span><span class="co"># get edge attributes</span></span>
<span><span class="fu">get_e</span><span class="op">(</span><span class="va">co_net</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##   id                         from                            to    weight</span></span>
<span><span class="co">## 1  1 s__un_f__Thermomonosporaceae     s__Actinocorallia_herbida 0.6759546</span></span>
<span><span class="co">## 2  2 s__un_f__Thermomonosporaceae       s__Kribbella_catacumbae 0.6742386</span></span>
<span><span class="co">## 3  3 s__un_f__Thermomonosporaceae       s__Kineosporia_rhamnosa 0.7378741</span></span>
<span><span class="co">## 4  4 s__un_f__Thermomonosporaceae   s__un_f__Micromonosporaceae 0.6236449</span></span>
<span><span class="co">## 5  5 s__un_f__Thermomonosporaceae s__Flavobacterium_saliperosum 0.6045747</span></span>
<span><span class="co">##         cor      p.value   e_type     width   color  e_class lty</span></span>
<span><span class="co">## 1 0.6759546 0.0020739524 positive 0.6759546 #48A4F0 e_class1   1</span></span>
<span><span class="co">## 2 0.6742386 0.0021502138 positive 0.6742386 #48A4F0 e_class1   1</span></span>
<span><span class="co">## 3 0.7378741 0.0004730567 positive 0.7378741 #48A4F0 e_class1   1</span></span>
<span><span class="co">## 4 0.6236449 0.0056818984 positive 0.6236449 #48A4F0 e_class1   1</span></span>
<span><span class="co">## 5 0.6045747 0.0078660171 positive 0.6045747 #48A4F0 e_class1   1</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And we can see, some attributes have been set when we build the network like <code>v_group</code>, these are internal attributes of metanet and are related to the following analysis and visualization.</p>
<div id="tbl-3-attribute" class="quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-3-attribute-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;3.1: Internal attributes of a <code>metanet</code>
</figcaption><div aria-describedby="tbl-3-attribute-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="table">
<colgroup>
<col style="width: 20%">
<col style="width: 79%">
</colgroup>
<thead><tr class="header">
<th>Attribute name</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>v_group</td>
<td>the big group for network, usually one omics data produces one group. Related to the vertex shape.</td>
</tr>
<tr class="even">
<td>v_class</td>
<td>some annotation of vertex, maybe classification or network modules. Related to the vertex color.</td>
</tr>
<tr class="odd">
<td>size</td>
<td>numeric variable assign to the vertex size.</td>
</tr>
<tr class="even">
<td>e_type</td>
<td>the type of edges, often be the positive or negative according to correlation. Related to the edge color.</td>
</tr>
<tr class="odd">
<td>width</td>
<td>numeric variable assign to the edge width.</td>
</tr>
<tr class="even">
<td>e_class</td>
<td>the second type of edges, often be the intra or inter according to two vertex group. Related to the edge line type.</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>We will talk about how to set these attributes for some specific analysis later.</p>
</section><section id="annotation" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="annotation">
<span class="header-section-number">3.2</span> Annotation</h2>
<p>Sometimes we have lots of annotation tables need to add to the network, such as abundance table, taxonomy table and so on, we can use <code>c_net_annotate()</code> to do this.</p>
<p>The annotation dataframe needs have rowname or a “name” column which match the vertex name of <code>metanet</code>, <code>c_net_annotate(mode = "v")</code> or <code>anno_vertex()</code> will automatically match the vertex name and combine the table.</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_annotate</span><span class="op">(</span><span class="va">co_net</span>, <span class="va">taxonomy</span><span class="op">[</span><span class="st">"Phylum"</span><span class="op">]</span>, mode <span class="op">=</span> <span class="st">"v"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net1</span></span>
<span><span class="fu">get_v</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##                           name  v_group  v_class size</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae v_group1 v_class1    4</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae v_group1 v_class1    4</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae v_group1 v_class1    4</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae v_group1 v_class1    4</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter v_group1 v_class1    4</span></span>
<span><span class="co">##                          label  shape   color            Phylum</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae circle #a6bce3 p__Actinobacteria</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae circle #a6bce3 p__Proteobacteria</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae circle #a6bce3 p__Proteobacteria</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae circle #a6bce3  p__Bacteroidetes</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter circle #a6bce3 p__Proteobacteria</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>c_net_annotate(mode = "e")</code> or <code>anno_edge()</code> receives the same format annotation dataframe, it will automatically match the “from” and “to” columns so that you can summary the links.</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"from"</span> <span class="op">=</span> <span class="st">"s__un_f__Thermomonosporaceae"</span>, <span class="st">"to"</span> <span class="op">=</span> <span class="st">"s__Actinocorallia_herbida"</span>, new_atr <span class="op">=</span> <span class="st">"new"</span><span class="op">)</span></span>
<span><span class="fu">c_net_annotate</span><span class="op">(</span><span class="va">co_net</span>, <span class="va">anno</span>, mode <span class="op">=</span> <span class="st">"e"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net1</span></span>
<span><span class="fu">get_e</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##   id                         from                            to    weight</span></span>
<span><span class="co">## 1  1 s__un_f__Thermomonosporaceae     s__Actinocorallia_herbida 0.6759546</span></span>
<span><span class="co">## 2  2 s__un_f__Thermomonosporaceae       s__Kribbella_catacumbae 0.6742386</span></span>
<span><span class="co">## 3  3 s__un_f__Thermomonosporaceae       s__Kineosporia_rhamnosa 0.7378741</span></span>
<span><span class="co">## 4  4 s__un_f__Thermomonosporaceae   s__un_f__Micromonosporaceae 0.6236449</span></span>
<span><span class="co">## 5  5 s__un_f__Thermomonosporaceae s__Flavobacterium_saliperosum 0.6045747</span></span>
<span><span class="co">##         cor      p.value   e_type     width   color  e_class lty new_atr</span></span>
<span><span class="co">## 1 0.6759546 0.0020739524 positive 0.6759546 #48A4F0 e_class1   1     new</span></span>
<span><span class="co">## 2 0.6742386 0.0021502138 positive 0.6742386 #48A4F0 e_class1   1    &lt;NA&gt;</span></span>
<span><span class="co">## 3 0.7378741 0.0004730567 positive 0.7378741 #48A4F0 e_class1   1    &lt;NA&gt;</span></span>
<span><span class="co">## 4 0.6236449 0.0056818984 positive 0.6236449 #48A4F0 e_class1   1    &lt;NA&gt;</span></span>
<span><span class="co">## 5 0.6045747 0.0078660171 positive 0.6045747 #48A4F0 e_class1   1    &lt;NA&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>MetaNet</code> provides function <code>c_net_set()</code> for easily annotation when you have more than one annotation table (it’s normal when you do a <strong>multi-omics analysis</strong>).</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Abundance_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"Abundance"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">totu</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># two annotation tables for vertex at same time, the rownames of Abundance_df and taxonomy do not need to be the same.</span></span>
<span><span class="va">co_net1</span> <span class="op">&lt;-</span> <span class="fu">c_net_set</span><span class="op">(</span><span class="va">co_net</span>, <span class="va">taxonomy</span><span class="op">[</span><span class="st">"Phylum"</span><span class="op">]</span>, <span class="va">Abundance_df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">get_v</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##                           name  v_group  v_class size</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae v_group1 v_class1    4</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae v_group1 v_class1    4</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae v_group1 v_class1    4</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae v_group1 v_class1    4</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter v_group1 v_class1    4</span></span>
<span><span class="co">##                          label  shape   color            Phylum Abundance</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae circle #a6bce3 p__Actinobacteria     26147</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae circle #a6bce3 p__Proteobacteria     25217</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae circle #a6bce3 p__Proteobacteria     16592</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae circle #a6bce3  p__Bacteroidetes     16484</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter circle #a6bce3 p__Proteobacteria     13895</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you have a vector and you absolutely know it matches the vertex name of network, you can use <code>igraph</code> method to annotate (Don’t recommend), same as the edge annotate vector. Refer to the <a href="https://r.igraph.org/">igraph manual</a> for more details.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">co_net1</span> <span class="op">&lt;-</span> <span class="va">co_net</span></span>
<span></span>
<span><span class="co"># add vertex attribute</span></span>
<span><span class="fu">V</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span><span class="op">$</span><span class="va">new_attri</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">get_v</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##                           name  v_group  v_class size</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae v_group1 v_class1    4</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae v_group1 v_class1    4</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae v_group1 v_class1    4</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae v_group1 v_class1    4</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter v_group1 v_class1    4</span></span>
<span><span class="co">##                          label  shape   color new_attri</span></span>
<span><span class="co">## 1 s__un_f__Thermomonosporaceae circle #a6bce3         1</span></span>
<span><span class="co">## 2        s__Pelomonas_puraquae circle #a6bce3         2</span></span>
<span><span class="co">## 3     s__Rhizobacter_bergeniae circle #a6bce3         3</span></span>
<span><span class="co">## 4     s__Flavobacterium_terrae circle #a6bce3         4</span></span>
<span><span class="co">## 5         s__un_g__Rhizobacter circle #a6bce3         5</span></span>
<span></span>
<span><span class="co"># add edge attribute</span></span>
<span><span class="fu">E</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span><span class="op">$</span><span class="va">new_attri</span> <span class="op">&lt;-</span> <span class="st">"new attribute"</span></span>
<span><span class="fu">get_e</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">##   id                         from                            to    weight</span></span>
<span><span class="co">## 1  1 s__un_f__Thermomonosporaceae     s__Actinocorallia_herbida 0.6759546</span></span>
<span><span class="co">## 2  2 s__un_f__Thermomonosporaceae       s__Kribbella_catacumbae 0.6742386</span></span>
<span><span class="co">## 3  3 s__un_f__Thermomonosporaceae       s__Kineosporia_rhamnosa 0.7378741</span></span>
<span><span class="co">## 4  4 s__un_f__Thermomonosporaceae   s__un_f__Micromonosporaceae 0.6236449</span></span>
<span><span class="co">## 5  5 s__un_f__Thermomonosporaceae s__Flavobacterium_saliperosum 0.6045747</span></span>
<span><span class="co">##         cor      p.value   e_type     width   color  e_class lty     new_attri</span></span>
<span><span class="co">## 1 0.6759546 0.0020739524 positive 0.6759546 #48A4F0 e_class1   1 new attribute</span></span>
<span><span class="co">## 2 0.6742386 0.0021502138 positive 0.6742386 #48A4F0 e_class1   1 new attribute</span></span>
<span><span class="co">## 3 0.7378741 0.0004730567 positive 0.7378741 #48A4F0 e_class1   1 new attribute</span></span>
<span><span class="co">## 4 0.6236449 0.0056818984 positive 0.6236449 #48A4F0 e_class1   1 new attribute</span></span>
<span><span class="co">## 5 0.6045747 0.0078660171 positive 0.6045747 #48A4F0 e_class1   1 new attribute</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="set-attributes" class="level2" data-number="3.3"><h2 data-number="3.3" class="anchored" data-anchor-id="set-attributes">
<span class="header-section-number">3.3</span> Set attributes</h2>
<p>As we mentioned before, there are some internal attributes have been set when building network, which are related to the network visualization (<a href="#tbl-3-attribute" class="quarto-xref">Table&nbsp;<span>3.1</span></a>). So we can use <code>c_net_set()</code> to custom these attributes fitting to our research needs.</p>
<p>Give the network various annotate tables and determine which columns use to set, we can give the columns name (one or more) to <code>vertex_group</code>, <code>vertex_class</code>, <code>vertex_size</code>, <code>edge_type</code>, <code>edge_class</code>, <code>edge_width</code>.</p>
<p>And the colors, linetypes, shapes and legends will be assigned automatically, just use <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code> to get the basic metanet figure. For more details to customize the plot, refer to the next section: <a href="04-visualization.html" class="quarto-xref"><span>Chapter 4</span></a>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"multi_test"</span>, package <span class="op">=</span> <span class="st">"MetaNet"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"c_net"</span>, package <span class="op">=</span> <span class="st">"MetaNet"</span><span class="op">)</span></span>
<span><span class="co"># build a multi-network</span></span>
<span><span class="va">multi1</span> <span class="op">&lt;-</span> <span class="fu">multi_net_build</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Microbiome <span class="op">=</span> <span class="va">micro</span>, Metabolome <span class="op">=</span> <span class="va">metab</span>, Transcriptome <span class="op">=</span> <span class="va">transc</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">multi1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set vertex_class</span></span>
<span><span class="va">multi1_with_anno</span> <span class="op">&lt;-</span> <span class="fu">c_net_set</span><span class="op">(</span><span class="va">multi1</span>, <span class="va">micro_g</span>, <span class="va">metab_g</span>, <span class="va">transc_g</span>, vertex_class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"kingdom"</span>, <span class="st">"type"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># set vertex_size</span></span>
<span><span class="va">multi1_with_anno</span> <span class="op">&lt;-</span> <span class="fu">c_net_set</span><span class="op">(</span><span class="va">multi1_with_anno</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"Abundance1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">micro</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"Abundance2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">metab</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"Abundance3"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">transc</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  vertex_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Abundance"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">multi1_with_anno</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-3-basic" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-3-basic-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-3-basic" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-3-basic-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-3-basic-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-3-basic-1.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-3-basic">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-3-basic-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Without annotation
</figcaption></figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-3-basic" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-3-basic-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-3-basic-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-3-basic-2.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-3-basic">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-3-basic-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) With annotation
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3-basic-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.1: Basic network plot
</figcaption></figure>
</div>
</section><section id="filter-sub-net" class="level2" data-number="3.4"><h2 data-number="3.4" class="anchored" data-anchor-id="filter-sub-net">
<span class="header-section-number">3.4</span> Filter (Sub-net)</h2>
<p>After setting your network properly, you may need to analysis part of the whole network (especially in the multi-omics analysis), <code>c_net_filter()</code> can get the sub-net conveniently, you can put lots of filter conditions (like the <code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter</a></code>) and get the sub-net you want.</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"multi_net"</span>, package <span class="op">=</span> <span class="st">"MetaNet"</span><span class="op">)</span></span>
<span><span class="va">multi2</span> <span class="op">&lt;-</span> <span class="fu">c_net_filter</span><span class="op">(</span><span class="va">multi1_with_anno</span>, <span class="va">v_group</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Microbiome"</span>, <span class="st">"Metabolome"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">c_net_filter</span><span class="op">(</span><span class="va">.</span>, <span class="va">e_class</span> <span class="op">==</span> <span class="st">"intra"</span>, mode <span class="op">=</span> <span class="st">"e"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">multi1_with_anno</span>, lty_legend <span class="op">=</span> <span class="cn">T</span>, main <span class="op">=</span> <span class="st">"Before filtering network"</span><span class="op">)</span> <span class="co"># before filter</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">multi2</span>, lty_legend <span class="op">=</span> <span class="cn">T</span>, main <span class="op">=</span> <span class="st">"After filtering network"</span><span class="op">)</span> <span class="co"># after filter</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-3-filter" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-3-filter-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-3-filter-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3-filter-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.2: Filter network
</figcaption></figure>
</div>
</div>
</div>
</section><section id="union-combine" class="level2" data-number="3.5"><h2 data-number="3.5" class="anchored" data-anchor-id="union-combine">
<span class="header-section-number">3.5</span> Union (Combine)</h2>
<p>If you have two networks and want to combine them, you can use <code>c_net_union()</code>.</p>
<div class="cell" data-layout="[[50,50],[100]]">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"c_net"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">co_net</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">co_net2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">co_net_union</span> <span class="op">&lt;-</span> <span class="fu">c_net_union</span><span class="op">(</span><span class="va">co_net</span>, <span class="va">co_net2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">co_net_union</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-5-union" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-5-union-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-5-union" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-5-union-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-5-union-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-5-union-1.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-5-union">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-5-union-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) network 1
</figcaption></figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-5-union" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-5-union-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-5-union-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-5-union-2.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-5-union">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-5-union-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) network 2
</figcaption></figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="cell-output-display quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-5-union" style="flex-basis: 100.0%;justify-content: flex-start;">
<div id="fig-5-union-3" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-5-union-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-5-union-3.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-5-union">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-5-union-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(c) union network
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-5-union-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.3: Union network
</figcaption></figure>
</div>
</div>
</section><section id="skeleton" class="level2" data-number="3.6"><h2 data-number="3.6" class="anchored" data-anchor-id="skeleton">
<span class="header-section-number">3.6</span> Skeleton</h2>
<p>If you want to summary the edges source and target according to one group, <code>summ_2col</code> is a easy way to get this information.</p>
<p><code>direct = F</code> argument means this is a undirected relationship, so “a-b” and “b-a” will summary to one type edge.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_annotate</span><span class="op">(</span><span class="va">co_net</span>, <span class="fu">select</span><span class="op">(</span><span class="va">taxonomy</span>, <span class="st">"Phylum"</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">"e"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net1</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">get_e</span><span class="op">(</span><span class="va">co_net1</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Phylum_from"</span>, <span class="st">"Phylum_to"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">summ_2col</span><span class="op">(</span><span class="va">df</span>, direct <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">count</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">Phylum_from_to</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">kbl</span><span class="op">(</span><span class="va">Phylum_from_to</span>, caption <span class="op">=</span> <span class="st">"Summary of edges according to Phylum"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">kable_paper</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">scroll_box</span><span class="op">(</span>width <span class="op">=</span> <span class="st">"100%"</span>, height <span class="op">=</span> <span class="st">"400px"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 0px; overflow-y: scroll; height:400px; overflow-x: scroll; width:100%; ">
<table class="lightable-paper table table-sm table-striped small" data-quarto-postprocess="true">
<caption>Summary of edges according to Phylum</caption>
<thead><tr class="header">
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">Phylum_from</th>
<th data-quarto-table-cell-role="th" style="text-align: left; position: sticky; top: 0; background-color: #FFFFFF;">Phylum_to</th>
<th data-quarto-table-cell-role="th" style="text-align: right; position: sticky; top: 0; background-color: #FFFFFF;">count</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">210</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">161</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">63</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: right;">62</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">54</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: right;">34</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: right;">25</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Acidobacteria</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: left;">p__Verrucomicrobia</td>
<td style="text-align: right;">13</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Chloroflexi</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Chloroflexi</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Planctomycetes</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Acidobacteria</td>
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Verrucomicrobia</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Chlamydiae</td>
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Chloroflexi</td>
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: left;">p__Verrucomicrobia</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Planctomycetes</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Proteobacteria</td>
<td style="text-align: left;">p__Spirochaetes</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Acidobacteria</td>
<td style="text-align: left;">p__Chloroflexi</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Chlamydiae</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Actinobacteria</td>
<td style="text-align: left;">p__Spirochaetes</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Chloroflexi</td>
<td style="text-align: left;">p__Verrucomicrobia</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Acidobacteria</td>
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Acidobacteria</td>
<td style="text-align: left;">p__Verrucomicrobia</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: left;">p__Chloroflexi</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Bacteroidetes</td>
<td style="text-align: left;">p__Spirochaetes</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Chlamydiae</td>
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Chlamydiae</td>
<td style="text-align: left;">p__Spirochaetes</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: left;">p__Planctomycetes</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">p__Firmicutes</td>
<td style="text-align: left;">p__Verrucomicrobia</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
<p>Then you can use sankey plot to display the links.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">pcutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/my_sankey.html">my_sankey</a></span><span class="op">(</span><span class="va">Phylum_from_to</span>, dragY <span class="op">=</span> <span class="cn">T</span>, fontSize <span class="op">=</span> <span class="fl">10</span>, width <span class="op">=</span> <span class="fl">600</span>, numberFormat <span class="op">=</span> <span class="st">",.4"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-3-sankey" class="cell-output-display quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-3-sankey-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sankeyNetwork html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-e08362da915e93ff5b62" style="width:100%;height:384px;"></div>
<script type="application/json" data-for="htmlwidget-e08362da915e93ff5b62">{"x":{"links":{"source":[0,0,0,0,0,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,3,3,3,4,4,4,5,5,5,5,6,7,7,7],"target":[8,9,10,11,12,8,13,14,9,10,15,11,16,12,13,9,10,11,16,12,10,11,16,10,11,12,10,15,11,12,11,11,16,12],"value":[6,2,1,21,1,62,25,2,8,34,3,161,2,5,5,1,11,54,1,4,1,5,1,5,9,2,10,1,63,1,7,210,3,13],"group":["p__Acidobacteria","p__Acidobacteria","p__Acidobacteria","p__Acidobacteria","p__Acidobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Actinobacteria","p__Bacteroidetes","p__Bacteroidetes","p__Bacteroidetes","p__Bacteroidetes","p__Bacteroidetes","p__Bacteroidetes","p__Chlamydiae","p__Chlamydiae","p__Chlamydiae","p__Chloroflexi","p__Chloroflexi","p__Chloroflexi","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Firmicutes","p__Planctomycetes","p__Proteobacteria","p__Proteobacteria","p__Proteobacteria"]},"nodes":{"name":["p__Acidobacteria","p__Actinobacteria","p__Bacteroidetes","p__Chlamydiae","p__Chloroflexi","p__Firmicutes","p__Planctomycetes","p__Proteobacteria","p__Actinobacteria ","p__Chloroflexi ","p__Firmicutes ","p__Proteobacteria ","p__Verrucomicrobia ","p__Bacteroidetes ","p__Chlamydiae ","p__Planctomycetes ","p__Spirochaetes "],"group":["p__Acidobacteria","p__Actinobacteria","p__Bacteroidetes","p__Chlamydiae","p__Chloroflexi","p__Firmicutes","p__Planctomycetes","p__Proteobacteria","p__Actinobacteria ","p__Chloroflexi ","p__Firmicutes ","p__Proteobacteria ","p__Verrucomicrobia ","p__Bacteroidetes ","p__Chlamydiae ","p__Planctomycetes ","p__Spirochaetes "],"posX":[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1]},"options":{"NodeID":"name","NodeGroup":"name","LinkGroup":"source","colourScale":"d3.scaleOrdinal().range(d3.schemeCategory20)","fontSize":12,"fontFamily":"arial","fontColor":null,"nodeWidth":15,"nodePadding":10,"nodeStrokeWidth":1,"nodeCornerRadius":5,"dragX":false,"dragY":true,"numberFormat":"pavian","orderByPath":true,"units":"","margin":{"top":null,"right":null,"bottom":null,"left":null},"iterations":1000,"zoom":false,"linkType":"bezier","curvature":0.5,"highlightChildLinks":true,"doubleclickTogglesChildren":true,"showNodeValues":true,"align":"none","xAxisDomain":["Phylum_from","Phylum_to"],"title":null,"nodeLabelMargin":2,"linkColor":"#A0A0A0","linkOpacity":0.5,"linkGradient":true,"nodeShadow":true,"scaleNodeBreadthsByString":true,"xScalingFactor":1,"yOrderComparator":null}},"evals":[],"jsHooks":[]}</script>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3-sankey-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.4: Sankey plot display edges according to Phylum
</figcaption></figure>
</div>
</div>
<p>Or use the circlize plot to display. <code>link_stat()</code> have more arguments than <code>summ_2col()</code> for summary edges.</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_set</span><span class="op">(</span><span class="va">co_net</span>, <span class="fu">select</span><span class="op">(</span><span class="va">taxonomy</span>, <span class="st">"Phylum"</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">co_net1</span></span>
<span><span class="fu">links_stat</span><span class="op">(</span><span class="va">co_net1</span>, topN <span class="op">=</span> <span class="fl">5</span>, group <span class="op">=</span> <span class="st">"Phylum"</span>, e_type <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-3-circlize" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-3-circlize-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-3-circlize-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3-circlize-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.5: Circlize plot display edges according to Phylum
</figcaption></figure>
</div>
</div>
</div>
<p>Other important action for a network is extracting the skeleton according to one group, which I call get skeleton plot. <code>get_group_skeleton()</code> can annotate all nodes with a group then combine each group nodes as a big node, and calculate new links between all groups.</p>
<p>This plot shows the flows between different groups clearly. By the way, the result will display by different edge types, if you just want summary all edges, set the e_type as one same character (use <code>c_net_set()</code>).</p>
<div class="cell">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">get_group_skeleton</span><span class="op">(</span><span class="va">co_net1</span>, Group <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">ske_net</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">ske_net</span>, vertex.label <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-3-skeleton-1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-3-skeleton-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-3-skeleton-1.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3-skeleton-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.6: Skeleton plot according to the Phylum
</figcaption></figure>
</div>
</div>
<div class="cell-output-display">
<div id="fig-3-skeleton-2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-3-skeleton-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="03-manipulation_files/figure-html/fig-3-skeleton-2.png" class="img-fluid figure-img" style="width:100.0%">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-3-skeleton-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3.7: Skeleton plot according to the Phylum
</figcaption></figure>
</div>
</div>
</div>
</section><section id="save" class="level2" data-number="3.7"><h2 data-number="3.7" class="anchored" data-anchor-id="save">
<span class="header-section-number">3.7</span> Save</h2>
<p>MetaNet also support to export various format network files (data.frame, graphml, pajek, lgl, …) for following analysis in other softwares (Cytoscape, Gephi…). You can use <code>c_net_save()</code> to save the network as a file, and use <code>c_net_load()</code> to load the network from a file.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">c_net_save</span><span class="op">(</span><span class="va">co_net</span>, filename <span class="op">=</span> <span class="st">"My_net"</span>, format <span class="op">=</span> <span class="st">"data.frame"</span><span class="op">)</span></span>
<span><span class="fu">c_net_save</span><span class="op">(</span><span class="va">co_net</span>, filename <span class="op">=</span> <span class="st">"My_net"</span>, format <span class="op">=</span> <span class="st">"graphml"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">c_net_load</span><span class="op">(</span><span class="st">"My_net.graphml"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./02-construction.html" class="pagination-link  aria-label=">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Construction</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./04-visualization.html" class="pagination-link" aria-label="<span class='chapter-number'>4</span>&nbsp; <span class='chapter-title'>Visualization</span>">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Visualization</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/Asa12138/MetaNet_q_book/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/Asa12138/MetaNet_q_book/blob/main/03-manipulation.qmd" class="toc-action"><i class="bi empty"></i>View source</a></li></ul></div></div></div></footer></body></html>